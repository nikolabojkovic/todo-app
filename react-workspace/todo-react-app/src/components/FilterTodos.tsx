import { Form, Stack } from "react-bootstrap";
import { useTodoList, useTodoListDispatch } from "../context/TodosContext";

export function FilterTodos() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function onFilter(completed: any, uncompleted: any) {
    dispatch({
      type: 'filtered',
      activePage: 1,
      filter: { 
        completed: completed,
        uncompleted: uncompleted
      }
    });
  }

  return (
    <Form className="App__filter d-flex justify-content-start align-items-center">
      <Stack direction="horizontal" gap={3}>       
        <Form.Check 
          type="switch"
          id={`completed`}
          label={`Completed`}
          checked={todoList.filter.completed}
          onChange={(e) => {
            onFilter(e.target.checked, todoList.filter.uncompleted);
          }}
        />   
        <Form.Check 
          type="switch"
          id={`uncompleted`}
          label={`Uncompleted`}
          checked={todoList.filter.uncompleted}
          onChange={(e) => {
            onFilter(todoList.filter.completed, e.target.checked);
          }}
        />         
      </Stack>
    </Form>
  )
}