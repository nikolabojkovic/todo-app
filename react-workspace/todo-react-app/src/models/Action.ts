export interface IAction {
  type: string,
  payload: any
}

export enum TodoActions {
  fetch = 'fetch',
  filter = 'filter',
  search = 'search',
  sort = 'sort',

  fetched = 'fetched',
  filtered = 'filtered',
  sorted = 'sorted',
  searched = 'searched',
  imported = 'imported',
  loadingStarted = 'loading-started',

  searchTermUpdated = 'searchTerm-updated',
  pagingUpdated = 'paging-updated',

  added = 'added',
  changed = 'changed',
  deleted = 'deleted',
}