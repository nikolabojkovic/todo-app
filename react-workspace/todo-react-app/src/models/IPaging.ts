
export interface IPaging {
	totalCount: number;
	activePage: number;
	itemsPerPage: number;
	startIndex: number;
	endIndex: number;
}

export interface IRange {
	offset: number;
	take: number;
}
