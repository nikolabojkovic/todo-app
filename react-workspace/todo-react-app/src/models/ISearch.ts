export interface ISearch {
	searchTerm: string;
}