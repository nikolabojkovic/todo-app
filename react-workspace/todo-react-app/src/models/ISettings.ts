export interface ISettings {
	general: IGeneralSettings;
	search: ISearchSettings;
	pagination: IPaginationSettings;
	theme: IThemeSettings;
}

export interface IGeneralSettings {
	isConfirmEnabled: boolean;
	isPaginationEnabled: boolean;
	isInfiniteScrollEnabled: boolean;
	listSizeType: ListContainerType;
	fixedListSize: number;
}

export interface ISearchSettings {
	isSearchOnKeyPressEnabled: boolean;
	debounceTime: number;
}

export interface IPaginationSettings {
	paginationType: PaginationType;
	maxVisiblePages: number;
}

export interface IThemeSettings {
	backgroundColor: BackgroundColor;
	primaryColor: string;
	bsTheme: BsThemes;
	primaryColorTopCord: number,
	primaryColorLefCord: number
}

export enum PaginationType {
	Rotate = 'Rotate',
	Classic = 'Classic'
}

export enum ListContainerType {
	Fixed = 'Fixed',
	Dynamic = 'Dynamic'
}

export enum BackgroundColor {
	DarkGray = 'dark-gray',
	DarkBlue = 'dark-blue',
	DarkRed = 'dark-red',
	LightGray = 'light-gray',
	LightBlue = 'light-blue',
	LightRed = 'light-red'
}

export enum BsThemes {
	Dark = 'dark',
	Light = 'light',
}

export const Themes = {
	dark: {
		gray: {
			backgroundPrimaryColor: "#282c34",
			backgroundSecondaryColor: "#444444",
			backgroundSecondaryColor1: "#4f4f4f",
			backgroundTernaryColor: "#5d5d5d",
			backgroundTernaryColor1: "#7e7d7d",
			backgroundTernaryColor2: "#6f6f6f",
		},
		blue: {
			backgroundPrimaryColor: "#282c34",
			backgroundSecondaryColor: "#0c1673",
			backgroundSecondaryColor1: "#0f1b8a",
			backgroundTernaryColor: "#111fa2",
			backgroundTernaryColor1: "#1324b9",
			backgroundTernaryColor2: "#1628d0",
		},
		red: {
			backgroundPrimaryColor: "#282c34",
			backgroundSecondaryColor: "#730c0c",
			backgroundSecondaryColor1: "#8a0f0f",
			backgroundTernaryColor: "#a21111",
			backgroundTernaryColor1: "#b91313",
			backgroundTernaryColor2: "#d01616",
		},
		
		textPrimaryColor: "white",
		textTernaryColor: "#767470",
		textTernaryColor2: "#A4A7AA",
	},
	light: {
		gray: {
			backgroundPrimaryColor: "white",
			backgroundSecondaryColor: "#ebebeb",
			backgroundSecondaryColor1: "#cfcfcf",
			backgroundTernaryColor: "#e0e0e0",
			backgroundTernaryColor1: "#cccccc",
			backgroundTernaryColor2: "#d3d2d2",
		},
		blue: {
			backgroundPrimaryColor: "white",
			backgroundSecondaryColor: "#eaecfa",
			backgroundSecondaryColor1: "#d5d9f6",
			backgroundTernaryColor: "#c1c5f1",
			backgroundTernaryColor1: "#acb2ec",
			backgroundTernaryColor2: "#979fe7",
		},
		red: {
			backgroundPrimaryColor: "white",
			backgroundSecondaryColor: "#faeaed",
			backgroundSecondaryColor1: "#f6d5db",
			backgroundTernaryColor: "#f1c1c9",
			backgroundTernaryColor1: "#ecacb8",
			backgroundTernaryColor2: "#e58e9e",
		},

		textPrimaryColor: "#282c34",
		textTernaryColor: "#A4A7AA",
		textTernaryColor2: "#313338",
	}
};