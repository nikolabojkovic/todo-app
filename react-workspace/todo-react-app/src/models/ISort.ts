export interface ISort {
	column: string;
	direction: SortDirection;
}

export enum SortDirection {
	None = 'none',
	Asc = 'asc',
	Desc = 'desc'
}