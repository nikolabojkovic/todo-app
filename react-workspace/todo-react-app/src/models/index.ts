export * from './Action';
export * from './Event';
export * from './IFilter';
export * from './IPaging';
export * from './ISearch';
export * from './ISettings';
export * from './ISort';
export * from './Todo';