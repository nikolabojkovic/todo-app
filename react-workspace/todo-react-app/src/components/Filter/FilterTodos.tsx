import { useState } from "react";
import { Form, Stack } from "react-bootstrap";
import Dropdown from 'react-bootstrap/Dropdown';
import { useTodoList, useTodoListDispatch } from "../../context/TodoListContext";
import { IFilter, StateFilter } from "../../models/IFilter";
import { IAction, TodoActions } from "../../models/Action";

type Props = { 
  filter: IFilter
}

export function FilterTodos({ filter }: Props) {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  // const [completedSwitch, setCompletedSwitch] = useState(completed);
  // const [uncompletedSwitch, setetUncompletedSwitch] = useState(uncompleted);
  const [stateFilter, setStateFilter] = useState(filter.state);

  function handleFilter(filterValue: StateFilter) {
    setStateFilter(filterValue);
    dispatch({
      type: TodoActions.filter,
      payload: {
        filter: { 
          state: filterValue
        } as IFilter, 
        sort: todoList.sort,
        searchTerm: todoList.search.searchTerm
      }
    } as IAction);
  }

  return (
    <Form className="App__filter d-flex justify-content-start align-items-center">
      <Stack direction="horizontal" gap={3}>       
        {/* <Form.Check 
          type="switch"
          id={`completed`}
          label={`Completed`}
          checked={completedSwitch}
          onChange={(e) => {
            setCompletedSwitch(e.target.checked);
            handleFilter(e.target.checked, uncompleted);
          }}
        />   
        <Form.Check 
          type="switch"
          id={`uncompleted`}
          label={`Uncompleted`}
          checked={uncompletedSwitch}
          onChange={(e) => {
            setetUncompletedSwitch(e.target.checked);
            handleFilter(completed, e.target.checked);
          }}
        /> */}
         <Dropdown>
            <Dropdown.Toggle
              data-testid="selected-filter-option"
              variant="outline-secondary"
              className="action-button ps-2 pe-2"
              id="dropdown-basic"
              disabled={todoList.isLoading}
            >
              {' '}{stateFilter}{' '}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item data-testid="filter-option-all" onClick={() => handleFilter(StateFilter.all)}>All</Dropdown.Item>
              <Dropdown.Item data-testid="filter-option-completed" onClick={() => handleFilter(StateFilter.completed)}>Completed</Dropdown.Item>
              <Dropdown.Item data-testid="filter-option-uncompleted" onClick={() => handleFilter(StateFilter.uncompleted)}>Uncompleted</Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
      </Stack>

    </Form>
  )
}