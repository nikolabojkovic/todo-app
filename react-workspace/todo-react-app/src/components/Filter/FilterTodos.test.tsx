import renderer from 'react-test-renderer';
import { FilterTodos } from './FilterTodos';
import { IFilter, StateFilter } from '../../models/IFilter';

describe('FilterTodos', () => {  
  it('component should match snapshot', () => {
    const jsxElement = <FilterTodos filter={{ state: StateFilter.all } as IFilter} />;
    const tree = renderer.create(jsxElement).toJSON();
    expect(tree).toMatchSnapshot();
  });
});