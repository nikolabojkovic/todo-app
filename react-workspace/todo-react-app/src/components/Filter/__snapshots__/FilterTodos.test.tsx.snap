// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterTodos component should match snapshot 1`] = `
<form
  className="App__filter d-flex justify-content-start align-items-center"
>
  <div
    className="hstack gap-3"
  >
    <div
      className="dropdown"
    >
      <button
        aria-expanded={false}
        className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
        disabled={false}
        id="dropdown-basic"
        onClick={[Function]}
        type="button"
      >
         
        All
         
      </button>
    </div>
  </div>
</form>
`;
